TabularClassifier.ipynb

```diff
@@ -0,0 +1,207 @@
+    "        )\n",
+    "    if not os.path.exists(TEST_FILE):\n",
+    "        os.system(\n",
+    "            f\"curl {CENSUS_INCOME_BASE_DOWNLOAD_URL}adult.test --output {TEST_FILE}\"\n",
+    "        )\n",
+    "\n",
+    "def reformat_test_csv():\n",
+    "    with open(TEST_FILE, \"r\") as fin:\n",
+    "        data = fin.read().splitlines(True)\n",
+    "    with open(TEST_FILE, \"w\") as fout:\n",
+    "        # for some reason each of the labels end with a \".\" in the test set\n",
+    "        # loop through data[1:] since the first line is bogus\n",
+    "        fout.writelines([line.replace(\".\", \"\") for line in data[1:]])\n",
+    "\n",
+    "def get_dataset_metadata():\n",
+    "    df = pd.read_csv(TEST_FILE)\n",
+    "    n_classes = df[df.columns[-1]].nunique()\n",
+    "    column_datatypes = []\n",
+    "    for col_type in df.dtypes[:-1]:\n",
+    "        if col_type == \"int64\":\n",
+    "            column_datatypes.append(\"numeric\")\n",
+    "        elif col_type == \"object\":\n",
+    "            column_datatypes.append(\"categorical\")\n",
+    "\n",
+    "    # the last column is the label\n",
+    "    column_datatypes.append(\"label\")\n",
+    "\n",
+    "    # TabularClassifier assumes no header so we add the first label \n",
+    "    # (considered a column name in pandas) as part of the test labels\n",
+    "    test_labels = [df.columns[-1]] + list(df[df.columns[-1]])\n",
+    "\n",
+    "    return n_classes, column_datatypes, test_labels\n",
+    "\n",
+    "\n",
+    "download_census_income_dataset()\n",
+    "reformat_test_csv()\n",
+    "(n_classes, column_datatypes, test_labels) = get_dataset_metadata()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 1: Defining the TabularClassifier model\n",
+    "\n",
+    "### Arguments\n",
+    "- `model_size`: Controls how big the model is. Options are ‘small’, ‘medium’,  ‘large’, or a target model size in gigabytes, i.e. “4 Gb” or “4Gb”.\n",
+    "- `n_classes`: How many output classes are in the dataset."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "classifier = bolt.TabularClassifier(model_size=\"medium\", n_classes=n_classes)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Step 2: Training the TabularClassifier\n",
+    "\n",
+    "### Arguments\n",
+    "- `train_file`: The dataset to train on. The expected format is a csv file with comma separated columns. This csv file should not have a header. The TabularClassifier also does not attempt to shuffle the dataset, the user should shuffle the dataset before calling train if necessary.\n",
+    "- `column_datatypes`: List of column types. Should be a list of strings with the types of the columns, which are one of: \"numeric\", \"categorical\", or \"label\". There should be only one \"label\" column. \n",
```

Comment:
```suggestion
    "- `column_datatypes`: List of column types. Should be a list of strings with the types of the columns, which are one of: \"numeric\", \"categorical\", or \"label\". There must be one and only \"label\" column. \n",
```

Reply:
can you explain what each column datatype should be? e.g. numeric is signed integer (floats?), categorical is arbitrary strings (one word? no commas? quoted?), label is ???